File: 3097-shortest-subarray-with-or-at-least-k-ii/3097-shortest-subarray-with-or-at-least-k-ii.cpp
class Solution {
public:
    void add(vector<int>& ones, int num){
        for(int k = 0; k < 32; k++){
            if(num & (1 << k)){
                ones[k]++;
            }
        }
    }
    
    void remove(vector<int>& ones, int num){
        for(int k = 0; k < 32; k++){
            if(num & (1 << k)){
                ones[k]--;
            }
        }
    }
    
    int getNum(vector<int>& ones){
        long long num = 0;
        for(int k = 0; k < 32; k++){
            if(ones[k]){
                num += (long long)pow(2, k);
            }
        }
        return num;
    }
    
    int minimumSubarrayLength(vector<int>& nums, int k) {
        int n = nums.size();
        
        vector<int> ones(32, 0);
        
        int minLen = INT_MAX;
        long long currNum;
        int i = 0, j = 0;
        while(j < n){
            add(ones, nums[j]);
            currNum = getNum(ones);
            while(currNum >= k && i <= j){
                minLen = min(minLen, (j - i + 1));
                remove(ones, nums[i]);
                currNum = getNum(ones);
                i++;
            }
            j++;
            
            cout<<endl;
        }
        
        return minLen == INT_MAX ? -1 : minLen;
    }
};

File: 3097-shortest-subarray-with-or-at-least-k-ii/3097-shortest-subarray-with-or-at-least-k-ii.java
class Solution {
    public int minimumSubarrayLength(int[] nums, int k) {
        int n = nums.length;
        
        int[] setBitCount = new int[32];
        
        int minLen = Integer.MAX_VALUE;
        int i = 0, j = 0;
        
        while(j < n) {
            addXorNum(setBitCount, nums[j]);
            
            while(getNum(setBitCount) >= k && i <= j) {
                minLen = Math.min(minLen, (j - i + 1));
                removeXorNum(setBitCount, nums[i]);
                i++;
            }
            
            j++;
        }
        
        return minLen == Integer.MAX_VALUE ? -1 : minLen;
    }
    
    private int getNum(int[] setBitCount) {
        int num = 0;
        
        for(int i = 0; i < 32; i++) {
            if(setBitCount[i] >= 1) {
                num = num | (1 << i);
            }
        }
        
        return num;
    }
    
    private void addXorNum(int[] setBitCount, int num) {
        for(int i = 0; i < 32; i++) {
            if(((1 << i) & num) != 0) {
                setBitCount[i]++;
            }
        }   
    }
    
    private void removeXorNum(int[] setBitCount, int num) {
        for(int i = 0; i < 32; i++) {
            if(((1 << i) & num) != 0 && setBitCount[i] > 0) {
                setBitCount[i]--;
            }
        }
    }
}

/*

    nums = [1,2,3], k = 2
    
    
    1 ->  000001
    2 ->  000010
    3 ->  000011
    
        
          000001 
          000010 <-
          000011
          
          
          000011
          
          000011 -> 3
*/